src/natcap/invest/seasonal_water_yield/seasonal_water_yield.py

```diff
@@ -879,133 +875,178 @@ def _calculate_annual_qfi(qfm_path_list, target_qf_path):
-        if n_events_nodata is not None:
-            valid_mask &= ~utils.array_equals_nodata(n_events, n_events_nodata)
-        # stream_nodata is the only input that carry over nodata values from
+        valid_p_mask = ~utils.array_equals_nodata(p_im, p_nodata)
+        valid_n_mask = ~utils.array_equals_nodata(n_m, n_nodata)
+        # precip mask: both p_im and n_m are defined and greater than 0
+        precip_mask = valid_p_mask & valid_n_mask & (p_im > 0) & (n_m > 0)
+        stream_mask = stream == 1
+        # stream_nodata is the only input that carries over nodata values from
         # the aligned DEM.
-        if stream_nodata is not None:
-            valid_mask &= ~utils.array_equals_nodata(
-                stream_array, stream_nodata)
-
-        valid_n_events = n_events[valid_mask]
-        valid_si = s_i[valid_mask]
+        valid_mask = (
+          valid_p_mask &
+          valid_n_mask &
+          ~utils.array_equals_nodata(stream, stream_nodata) &
+          ~utils.array_equals_nodata(s_i, si_nodata))
+
+        # QF is defined in terms of three cases:
+        #
+        # 1. Where there is no precipitation, QF = 0
+        #    (even if stream or s_i is undefined)
+        #
+        # 2. Where there is precipitation and we're on a stream, QF = P
+        #    (even if s_i is undefined)
+        #
+        # 3. Where there is precipitation and we're not on a stream, use the
+        #    quickflow equation (only if all four inputs are defined):
+        #    QF_im = 25.4 * n_m * (
+        #       (a_im - s_i) * exp(-0.2 * s_i / a_im) +
+        #       s_i^2 / a_im * exp(0.8 * s_i / a_im) * E1(s_i / a_im)
+        #    )
+        #
+        # When evaluating case 3, there are a couple of edge cases related
+        # to the last term of the quickflow equation:
+        #
+        #       s_i^2 / a_im * exp(0.8 * s_i / a_im) * E1(s_i / a_im)
+        #
+        # 3a. Where s_i = 0, this term should equal 0 (per conversation with
+        #     Rafa). But, evaluating in numpy you get NaN and a warning
+        #     because E1(0 / a_im) = infinity.
+        #
+        #     Solution: Preemptively set the result to 0 where s_i = 0 in
+        #     order to avoid calculations with infinity.
+        #
+        # 3b. When the ratio s_i / a_im becomes large, this term approaches 0.
+        #     [NOTE: I don't know how to prove this mathematically, but it
+        #     holds true when I tested with reasonable values of s_i and a_im].
+        #     The exp() term becomes very large, while the E1() term becomes
+        #     very small. When (s_i / a_im) > 880 ish, exp(0.8 * s_i / a_im)
+        #     will overflow and round up to infinity. Meanwhile, E1(s_i / a_im)
+        #     will become so small that it rounds down to 0.
+        #
+        #     Solution: Catch the overflow warning, and set the term to 0
+        #     anywhere that overflow happens.
```

Comment:
Values of `s_i / a_im` that make `exp` overflow aren't common, but are totally possible with reasonable input data, for example:
```
CN = 30
P = 8 (millimeters of rain per month)
n = 10 (number of rain events per month)

s_i = 1000/CN - 10 = 23.33
a_im = P/(n * 25.4) = 8/(10 * 25.4) = 0.031

s_i / a_im = 23.33 / 0.031 = 752

exp(752) = 3.8 * 10^326, which overflows float64.
```

as far as numerical constraints, `exp(x)` overflows when `x` is about 709.78. I'm not sure if this might vary on different systems.

Reply:
> We're setting to 0 because if the exp has overflowed, then we know the E1 term is 0?

yes, the E1 will be extremely close to 0 and "underflow" to 0. for example:
```
>>> scipy.special.exp1(200)
6.885226106307636e-90
>>> scipy.special.exp1(400)
4.776013586420972e-177
>>> scipy.special.exp1(700)
1.406518766234033e-307
>>> scipy.special.exp1(800)
0.0
```

Reply:
@emlys thanks for making this example. It makes numerically sense, but let's look at it from a process perspective. basically we have no precip (0.8 mm per rain is basically no rain) meeting a very permeable soil (very low curve number). As I see from the above, if we just use the threshold value of exp(709) then QF would be 0, right? Probably we can check that somehow bottom up? Calculate for each pixel what values of x would result in an overflow and set quickflow to 0 there? There  might e a more elegant way. Let's think about it after thanksgiving.